### DISCUSSION TITLE
Add support for PTX 5.0 ISA

### DISCUSSION THREAD

**Discussion by Hdksg10** [gpuocelot/gpuocelot#28] [General]
Hi there,
I'm working to add full PTX 5.0 ISA support for gpuocelot (as a PTX to LLVM translator) , and have implemented the [lop3](https://docs.nvidia.com/cuda/archive/8.0/parallel-thread-execution/index.html#logic-and-shift-instructions-lop3) instruction in my [fork](https://github.com/Hdksg10/gpuocelot/tree/pipeline-test).
However, I'm confused about how to write test for translated instruction in gpuocelot. Currently, I'm using JIT compiler both from cuda toolkit and gpuocelot to verify the translated llvm kernel and original ptx kernel have same result.
Any advice on it?

**Comment by dmikushin** [comment]
Hi @Hdksg10 , thank you for working on this!

I understand that currently you execute two versions of code and ensure results equality. This approach makes sense to me!

But creating a lot of tests by hand is a lot of work. One advise would be to look for an existing testing material and try to re-use it.

LLVM has a PTX backend, which should have a lot of IR->PTX code generation tests. What you could do is to _reverse_ those tests to ensure proper LLVM IR generation for the given PTX code. The generated code could have some variability, which is accepted by LLVM linter using regular expressions matching.
